primitives/src/merkle_tree/merkle_tree_traits.rs

```diff
@@ -0,0 +1,158 @@
+    Default
+    + Ord
+    + Clone
+    + Copy
+    + Debug
+    + CanonicalSerialize
+    + CanonicalDeserialize
+    + Eq
+    + PartialEq
+    + AsRef<[F]>
+{
+    /// Convert the element into a slice of field elements
+    /// TODO: is it the same as `AsRef<[F]>`?
+    fn as_slice_ref(&self) -> &[F];
+}
+
+/// Merkle tree hash function
+pub trait Hasher<F: Field> {
+    /// Digest a list of values
+    fn digest(data: &[F]) -> F;
+}
+
+/// Basic functionalities for a merkle tree implementation
+pub trait MerkleTree<F: Field>: Sized {
+    /// Merkle tree element type
+    type ElementType: ElementType<F>;
+    /// Hash algorithm used in merkle tree
+    type Hasher: Hasher<F>;
+    /// Leaf arity
+    type LeafArity: Unsigned;
+    /// Non-leaf arity
+    type TreeArity: Unsigned;
+    /// Exsistential proof
+    type Proof;
+    /// Batch proof
+    type BatchProof;
+
+    /// Construct a new merkle tree with given height from a data slice
+    fn build(
+        height: usize,
+        data: impl Iterator<Item = Self::ElementType>,
+    ) -> Result<Self, PrimitivesError>;
+
+    /// Return the height of this merkle tree
+    fn height(&self) -> usize;
+    /// Return the maximum allowed number leaves
+    fn capacity(&self) -> u64;
+    /// Return the current number of leaves
+    fn num_leaves(&self) -> u64;
+
+    /// Return the current root value
+    fn value(&self) -> F;
+
+    /// Returns the leaf value given a position
+    /// * `pos` - zero-based index of the leaf in the tree
+    /// * `returns` - Leaf value at the position. LookupResult::EmptyLeaf if the
+    ///   leaf position is empty or invalid, LookupResult::NotInMemory if the
+    ///   leaf position has been forgotten.
+    fn lookup(&self, pos: u64) -> LookupResult<(), Self::Proof>;
+
+    /// Verify an element is a leaf of a Merkle tree given the proof
+    /// * `pos` - zero-based index of the leaf in the tree
+    /// * `proof` - a merkle tree proof
+    /// * `returns` - Err() if something is wrong with this proof, Ok(result)
+    ///   otherwise
+    fn verify(&self, pos: u64, proof: impl Borrow<Self::Proof>) -> Result<bool, PrimitivesError>;
+
+    // fn batch_lookup(&self, pos: impl Iterator<Item = usize>) -> LookupResult<(),
+    // Self::BatchProof>; fn batch_verify(
+    //     &self,
+    //     pos: impl Iterator<Item = usize>,
+    //     proof: impl Borrow<Self::BatchProof>,
+    // ) -> Result<(), PrimitivesError>;
+}
+
+/// Merkle tree that allows insertion at back
+pub trait AppendableMerkleTree<F: Field>: MerkleTree<F> {
+    /// Insert a new value at the leftmost available slot
+    /// * `elem` - element to insert in the tree
+    /// * `returns` - Ok(()) if successful
+    fn push(&mut self, elem: &Self::ElementType) -> Result<(), PrimitivesError>;
+
+    /// Insert a list of new values at the leftmost available slots
+    /// * `elems` - elements to insert
+    /// * `returns` - Ok(()) if successful
+    fn emplace(
```

Comment:
Is there a reason we went with `emplace`? I think it's more idiomatic to call this `extend`. `emplace` reminds me of C++ emplacement `new`, but that's not what this is at all.

Reply:
No particular reason. Will return to `extend`.